### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by vandermesis**

### CODE REVIEW

File: CleanWeatherTests/Scenes/CityList/CityListPresenterTests.swift [hunk 1]
```diff
@@ -6,12 +6,33 @@
...
 
+    override func spec() {
+
+        var presenter: CityListPresenterImpl<FakeCityListController>!
+        var controller: FakeCityListController!
+
+        beforeEach {
+            presenter = CityListPresenterImpl<FakeCityListController>()
+            controller = FakeCityListController()
+        }
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityListPresenterTests.swift:25]: presenter.controller = controller

you forgot to set controller for presenter

File: CleanWeatherTests/Scenes/CityList/CityListRouterTests.swift [hunk 2]
```diff
@@ -6,12 +6,34 @@
...
+            router = CityListRouterImpl()
+            controller = CityListCreator().getController()
+            navigationController = UINavigationController()
+            navigationController.viewControllers = [controller]
+            router.controller = controller
+        }
+
+        //FIXME: router should not be nil here
+        describe("should navigate to details") {
+//            router.navigateToCityDetails(cityWeather: Mock.cityWeather1)
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityListRouterTests.swift]: need this in beforeEach closure

File: CleanWeatherTests/Scenes/PeopleList/PeopleListPresenterTests.swift [hunk 3]
```diff
@@ -36,12 +36,12 @@ final class PeopleListPresenterTests: QuickSpec {
...
             it("should have valid names") {
-                //TODO: Remove this comment after understanding:
-                // our presenter converts Person to PersonDisplayable, which is joining first and last name, that's why I check it here
-                // but if you want you can add more checks
-                // and maybe better if you create your own unit test for 'PersonDisplayable' where you give a normal 'Person' and you validates if all fields are set properly :)
                 expect(controller.displayPeopleCalled?[4].name).to(equal("Dael Vignaux"))
             }
+
+            it("should have valid genderIconName") {
+                expect(controller.displayPeopleCalled?[0].genderIconName).to(equal("femaleIcon"))
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/PeopleList/PeopleListPresenterTests.swift]: better to check something else than things related with UI, icon can change

File: CleanWeatherTests/Scenes/PeopleList/PersonDisplayableTests.swift [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import Quick
+import Nimble
+
+@testable import CleanWeather
+
+final class PersonDisplayableTests: QuickSpec {
+
+    override func spec() {
+
+        var person: [Person]!
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/PeopleList/PersonDisplayableTests.swift]: you're goal is to check PersonDisplayable model
so you create a file for that - thats ok
and you check in all describe closures all methods you want to test - in your example only custom Init

so you don't need an array of people. just test all parameters only for 1 object, no need arrays

File: CleanWeatherTests/Scenes/PeopleList/PersonDisplayableTests.swift [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+            it("should return a valid genderIconName") {
+                expect(displayable[1].genderIconName).to(equal("femaleIcon"))
+                expect(displayable[2].genderIconName).to(equal("maleIcon"))
+            }
+
+            it("should return all properties of type String") {
+                expect(displayable[2].id).to(beAKindOf(String.self))
+                expect(displayable[2].name).to(beAKindOf(String.self))
+                expect(displayable[2].age).to(beAKindOf(String.self))
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/PeopleList/PersonDisplayableTests.swift]: and do not check for kind of classes,. but actuall values, you can expect it to be "52" etc

**Review by SredzinskiPatryk — CHANGES_REQUESTED** [review]

File: CleanWeatherTests/Scenes/CityForecast/FakeCityForecastController.swift [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import UIKit
+
+@testable import CleanWeather
+
+final class FakeCityForecastController: UIViewController, CityForecastPresentable {
+
+    var displayCityDetailsCalled: CityForecastDisplayable?
+    var displayCityForecastCalled: [CityForecastListDisplayable]?
+
+
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityForecast/FakeCityForecastController.swift]: 2 entery

File: CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift [hunk 7]
```diff
@@ -6,12 +6,100 @@
...
+                    expect(presenter.presentAlertMessageCalled).to(beNil())
+                    expect(presenter.presentAlertTitleCalled).to(beNil())
+                    expect(presenter.presentErrorCalled).to(beNil())
+                }
+            }
+
+            context("on failure response") {
+
+                beforeEach {
+                    worker.fetchCityWeatherCompletion?(.failure(AppError(message: R.string.localizable.networkingError())))
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift]: uzywanie localizable stringow w teestach jest bledne, bo moze prowadzic do problemow jak np testy beda sie odpalac na jakims tam urzadzeniu ktory ma ustawiony inny jezyk

napisz sobie jakis wlasny pusty error:

struct UnitTestError: Error {} i go uzywaj

File: CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift [hunk 8]
```diff
@@ -6,12 +6,100 @@
...
+            router = FakeCityListRouter()
+            worker = FakeCityListWorker()
+            presenter = FakeCityListPresenter()
+            interactor = CityListInteractorImpl(presenter: presenter, worker: worker, router: router)
+        }
+
+        describe("getting City") {
+
+            beforeEach {
+                interactor.getCity()
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift]: getCity zla nazwa metody w interactorze
predzej getCitiesWeather

File: CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift [hunk 9]
```diff
@@ -6,12 +6,100 @@
...
+        }
+
+        describe("selecting city cell") {
+
+            beforeEach {
+                interactor.getCity()
+                worker.fetchCityWeatherCompletion?(.success(Mock.citiesWeather))
+            }
+
+            context("on success response") {
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift]: on valid id tapped

i dodaj on invalid id tapped

zeby sprawdzic ten case co ci sie teraz na czerwono swieci w iinteractorze ze nie prrzetestowane (ten guard let)

File: CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift [hunk 10]
```diff
@@ -6,12 +6,100 @@
...
+
+                beforeEach {
+                    interactor.didSelectCityCell(id: Mock.cityWeather4.id)
+                }
+
+                it("should filter [CityWeather] with valid id") {
+                    expect(router.navigateToCityDetailsCityWeatherCalled?.id).to(equal(Mock.cityWeather4.id))
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift]: ten test wywal, nastepny robi to samo

File: CleanWeatherTests/Scenes/CityList/CityListPresenterTests.swift [hunk 11]
```diff
@@ -6,12 +6,38 @@
...
+                presenter.displayCitiesWeather(citiesWeather: Mock.citiesWeather)
+            }
+
+            it("should call controller to display cities") {
+                expect(controller.displayCityCalled).notTo(beNil())
+                expect(controller.displayCityCalled?.count).to(equal(Mock.citiesWeather.count))
+            }
+
+            it("should display temperature with degree sign") {
+                expect(controller.displayCityCalled?[0].temp).to(contain("°"))
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityListPresenterTests.swift]: equal("26.4°") a nie contain, sprawdz cala tempeerature

File: CleanWeatherTests/Scenes/CityList/CityListWorkerTests.swift [hunk 12]
```diff
@@ -6,12 +6,78 @@
...
+
+                it("should return no errors") {
+                    expect(receivedError).to(beNil())
+                }
+            }
+
+            context("on error response") {
+
+                beforeEach {
+                    networking.fetchCurrentWeatherForAllCitiesCompletion?(.failure(AppError(message: R.string.localizable.apiError())))
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityListWorkerTests.swift]: uzyj wlasnego erroru bez stringow do testow

File: CleanWeatherTests/Scenes/CityList/CityWeatherDisplayableTests.swift [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+            it("should return valid id string") {
+                expect(displayable.id).to(equal("05197501-e7b0-4bfd-84e0-e07038270f2d"))
+            }
+
+            it("should return valid name string") {
+                expect(displayable.name).to(equal("Gdańsk"))
+            }
+
+            it("should return valid temperature string") {
+                expect(displayable.temp).to(equal("0°"))
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityWeatherDisplayableTests.swift]: lepszy bylby przyklad z temperatura inna niz 0

File: CleanWeatherTests/Scenes/CityList/CityWeatherDisplayableTests.swift [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+    override func spec() {
+
+        var displayable: CityWeatherDisplayable!
+
+        beforeEach {
+            displayable = CityWeatherDisplayable(object: Mock.cityWeather3)
+        }
+
+        describe("init of CityWeatherDisplayable") {
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityWeatherDisplayableTests.swift:20]: beforeeach w tym closure powinien byc

**Review by SredzinskiPatryk — CHANGES_REQUESTED** [review]

File: CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift (hunk 9 above)
Comment (vandermesis) [reply CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift]: Właśnie miałem pytać, czy testujemy to guard let, bo się świeci na czerwono. Nie byłem pewien, bo w sumie ten optional tam powstaje tylko przez to, że zastosowałem .first, gdybym użył [0] to by go nie było. W sumie mamy pewność, że ten array zwróci prawidłowy CityWeather.

File: CleanWeather/Extensions/DoubleExtension.swift [hunk 15]
```diff
@@ -15,6 +15,6 @@ extension Double {
     }
     
     var percentString: String {
-        return "\(Int(self.rounded()))%"
+        return "\(Int(self * 100))%"
```
Comment (SredzinskiPatryk) [inline CleanWeather/Extensions/DoubleExtension.swift]: return "\((self * 100).rounded())%"

wouldnt it be better?

File: CleanWeatherTests/Mocks/Mock+CityForecast.swift [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+//
+//  Created by Marek Skrzelowski on 27/11/2019.
+//  Copyright © 2019 vandermesis. All rights reserved.
+//
+
+@testable import CleanWeather
+
+extension Mock {
+
+    static let cityForecast1 = CityForecast(id: "c1f2934e-b2f4-47d2-961f-22c72d2411b5", hour: "12", hourTemp: 10.34, hourPrecipProbability: 0.46, icon: "clear-day")
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Mocks/Mock+CityForecast.swift]: do we need to have hour as a string? that limits some features that we could use, for example sorting

File: CleanWeatherTests/Scenes/CityForecast/CityForecastInteractorTests.swift [hunk 17]
```diff
@@ -6,12 +6,90 @@
...
+                }
+
+                it("should call presenter to hide spinner") {
+                    expect(presenter.toggleSpinnerStateCalled).to(beFalse())
+                }
+
+                it("should call presenter to display valid hourly city details list for valid id") {
+                    expect(presenter.displayCityDetailsListCalled).notTo(beNil())
+                    expect(presenter.displayCityDetailsListCalled?.count).to(equal(Mock.cityForecast.count))
+                    expect(Mock.cityWeather2.id).to(equal(Mock.cityForecast1.id))
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityForecast/CityForecastInteractorTests.swift]: this never changes, youre testing 2 mocked values, that makes no sense and is not related with fetching city hour details list youre now testing

File: CleanWeatherTests/Scenes/CityForecast/CityForecastWorkerTests.swift [hunk 18]
```diff
@@ -6,12 +6,79 @@
...
-class CityForecastWorkerTests: XCTestCase {
+final class CityForecastWorkerTests: QuickSpec {
 
+    override func spec() {
+
+        var networking: FakeWeatherNetoworking!
+        var worker: CityForecastWorkerImpl!
+
+        beforeEach {
+            networking = FakeWeatherNetoworking()
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityForecast/CityForecastWorkerTests.swift]: FakeWeatherNetoworking
FakeWeatherNetworking

File: CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift [hunk 19]
```diff
@@ -6,12 +6,107 @@
...
+        var interactor: CityListInteractorImpl!
+
+        beforeEach {
+            router = FakeCityListRouter()
+            worker = FakeCityListWorker()
+            presenter = FakeCityListPresenter()
+            interactor = CityListInteractorImpl(presenter: presenter, worker: worker, router: router)
+        }
+
+        describe("getting City") {
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Scenes/CityList/CityListInteractorTests.swift]: getting cities

**Review by SredzinskiPatryk — CHANGES_REQUESTED** [review]

File: CleanWeather/Extensions/DoubleExtension.swift (hunk 15 above)
Comment (vandermesis) [reply CleanWeather/Extensions/DoubleExtension.swift]: Why would that be better?
I would like to have 43% displayed for example - not 40%.

File: CleanWeather/Extensions/ErrorExtension.swift [hunk 20]
```diff
@@ -12,13 +12,18 @@ struct AppError: Error {
...
+struct UnitTestError: Error {}
+
 extension Error {
     var userFriendlyMessage: String {
         if self is DecodingError {
             return R.string.localizable.decodingError()
         } else if let appError = self as? AppError {
             return appError.message
+        } else if self is UnitTestError {
+            return R.string.localizable.unitTestError()
```
Comment (SredzinskiPatryk) [inline CleanWeather/Extensions/ErrorExtension.swift]: do not localize strings that aren't going to an app. Unit Test Error should be only in Test target

File: CleanWeather/Support/en.lproj/Localizable.strings [hunk 21]
```diff
@@ -11,6 +11,9 @@
 "unknownError" = "Unknown Error";
 "decodingError" = "Decoding Error";
 "dataSourceError" = "Cell DataSource Error";
+"networkingError" = "Networking Error";
+"apiError" = "API didn't response";
+"unitTestError" = "Unit Test Error";
```
Comment (SredzinskiPatryk) [inline CleanWeather/Support/en.lproj/Localizable.strings]: no need to translate something that is not in release

File: CleanWeatherTests/Extensions/ErrorExtensionTests.swift [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+
+@testable import CleanWeather
+
+final class ErrorExtensionTests: QuickSpec {
+
+    override func spec() {
+
+        var error: Error!
+
+        describe("error user friendly message") {
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Extensions/ErrorExtensionTests.swift]: userfriendlymessage is using localizable strings, it'll fail if you get other translations, this whole describe is pointless

File: CleanWeatherTests/Models/WeatherSymbolTests.swift [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+
+@testable import CleanWeather
+
+final class WeatherSymbolTests: QuickSpec {
+
+    override func spec() {
+
+        var weatherSymbol: WeatherSymbol!
+
+        describe("init of WeatherSymbol") {
```
Comment (SredzinskiPatryk) [inline CleanWeatherTests/Models/WeatherSymbolTests.swift]: checking for icon names is related to UI testing I'd remove it

**Review by SredzinskiPatryk — CHANGES_REQUESTED** [review]

File: CleanWeather/Extensions/ErrorExtension.swift [hunk 24]
```diff
@@ -12,13 +12,16 @@ struct AppError: Error {
     let message: String
 }
 
+struct UnitTestError: Error {}
```
Comment (SredzinskiPatryk) [inline CleanWeather/Extensions/ErrorExtension.swift]: przenies do Test tarrgeetu w jakims nowym pliku w katalogu z testami

**Review by SredzinskiPatryk — CHANGES_REQUESTED** [review]

**Pull request merged**